File: java/muse/data_structures/LinkedList.java
package muse.data_structures;

import muse.data_structures.interfaces.IList;

public class LinkedList<E> implements IList<E> {
    private static class Node<E> {
        public E data;
        public Node<E> next = null;

        public Node(E element) {
            data = element;
        }
    }

    private Node<E> head = null;
    private Node<E> tail = null;
    private int length = 0;

    @Override
    public int size() {
        return length;
    }

    @Override
    public boolean isEmpty() {
        return length == 0;
    }

    @Override
    public E get(int index) {
        if (index < 0 || index >= length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> cursor;

        if (index == length - 1) {
            cursor = tail;
        } else {
            cursor = head;
            for (int i = 0; i < index; i++) {
                cursor = cursor.next;
            }
        }

        return cursor.data;
    }

    @Override
    public void set(int index, E element) {
        if (index < 0 || index >= length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> cursor;

        if (index == length - 1) {
            cursor = tail;
        } else {
            cursor = head;
            for (int i = 0; i < index; i++) {
                cursor = cursor.next;
            }
        }

        cursor.data = element;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void insert(int index, E element) {
        if (index < 0 || index > length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> node = new Node<>(element);

        if (index == 0) {
            if (length != 0) {
                node.next = head;
            } else {
                tail = node;
            }
            head = node;
        } else if (index == length) {
            tail.next = node;
            tail = node;
        } else {
            Node<E> cursor = head;
            for (int i = 0, bound = index - 1; i < bound; i++) {
                cursor = cursor.next;
            }
            node.next = cursor.next;
            cursor.next = node;
        }

        length++;
    }

    @Override
    public void remove(int index) {
        if (index < 0 || index >= length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        Node<E> target;

        if (index == 0) {
            target = head;
            if (length == 1) {
                head = null;
                tail = null;
            } else {
                head = target.next;
            }
        } else {
            Node<E> cursor = head;
            for (int i = 0, bound = index - 1; i < bound; i++) {
                cursor = cursor.next;
            }
            target = cursor.next;
            cursor.next = target.next;
            if (index == length - 1) {
                tail = cursor;
            }
        }

        target.data = null;

        length--;
    }

    @Override
    public E front() {
        return get(0);
    }

    @Override
    public E back() {
        return get(length - 1);
    }

    @Override
    public void prepend(E element) {
        insert(0, element);
    }

    @Override
    public void append(E element) {
        insert(length, element);
    }

    @Override
    public void poll() {
        remove(0);
    }

    @Override
    public void eject() {
        remove(length - 1);
    }
}

File: php/muse/data_structures/LinkedList.php
<?php

declare(strict_types=1);

namespace muse\data_structures\LinkedList;

require_once "muse/data_structures/interfaces/IList.php";

use muse\data_structures\interfaces\IList;

class Node
{
    public $data;
    public $next = null;

    public function __construct(int $element)
    {
        $this->data = $element;
    }
}

class LinkedList implements IList\IList
{
    private $head = null;
    private $tail = null;
    private $length = 0;

    public function __construct()
    {
    }

    public function size(): int
    {
        return $this->length;
    }

    public function isEmpty(): bool
    {
        return $this->length == 0;
    }

    public function get(int $index): int
    {
        if ($index < 0 || $index >= $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $cursor = null;

        if ($index == $this->length - 1) {
            $cursor = $this->tail;
        } else {
            $cursor = $this->head;
            for ($i = 0; $i < $index; $i++) {
                $cursor = $cursor->next;
            }
        }

        return $cursor->data;
    }

    public function set(int $index, int $element): void
    {
        if ($index < 0 || $index >= $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $cursor = null;

        if ($index == $this->length - 1) {
            $cursor = $this->tail;
        } else {
            $cursor = $this->head;
            for ($i = 0; $i < $index; $i++) {
                $cursor = $cursor->next;
            }
        }

        $cursor->data = $element;
    }

    public function insert(int $index, int $element): void
    {
        if ($index < 0 || $index > $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $node = new Node($element);

        switch ($index) {
            case 0:
                if ($this->length != 0) {
                    $node->next = $this->head;
                } else {
                    $this->tail = $node;
                }
                $this->head = $node;
                break;
            case $this->length:
                $this->tail->next = $node;
                $this->tail = $node;
                break;
            default:
                $cursor = $this->head;
                for ($i = 0, $bound = $index - 1; $i < $bound; $i++) {
                    $cursor = $cursor->next;
                }
                $node->next = $cursor->next;
                $cursor->next = $node;
        }

        $this->length++;
    }

    public function remove(int $index): void
    {
        if ($index < 0 || $index >= $this->length) {
            throw new RuntimeException("[PANIC - IndexOutOfBounds]");
        }

        $target = null;

        if ($index == 0) {
            $target = $this->head;
            if ($this->length == 1) {
                $this->head = null;
                $this->tail = null;
            } else {
                $this->head = $target->next;
            }
        } else {
            $cursor = $this->head;
            for ($i = 0, $bound = $index - 1; $i < $bound; $i++) {
                $cursor = $cursor->next;
            }
            $target = $cursor->next;
            $cursor->next = $target->next;
            if ($index == $this->length - 1) {
                $this->tail = $cursor;
            }
        }

        $target->data = null;

        $this->length--;
    }

    public function front(): int
    {
        return $this->get(0);
    }

    public function back(): int
    {
        return $this->get($this->length - 1);
    }

    public function prepend(int $element): void
    {
        $this->insert(0, $element);
    }

    public function append(int $element): void
    {
        $this->insert($this->length, $element);
    }

    public function poll(): void
    {
        $this->remove(0);
    }

    public function eject(): void
    {
        $this->remove($this->length - 1);
    }
}
